enigma-runtime-t/src/lib.rs

```diff
@@ -209,34 +210,62 @@ impl Runtime {
         Ok(())
     }
 
+    fn treat_gas_overflow(&mut self, val: &Option<u64>) -> Result<()>{
+        if val.is_none() {
+            self.gas_counter = self.gas_limit;
+            Err(WasmError::GasLimit)
+        }
+        else {
+            Ok(())
+        }
+    }
+
+    fn treat_gas_underflow(&mut self, val: &Option<u64>) -> (bool, u64){
+        let mut result = 0;
+        if val.is_none() {
+            self.gas_return = self.gas_limit;
+            (true, result)
+        } else {
+            (false, val.unwrap())
+        }
+    }
+
     /// args:
    /// * `new_value_len` - the length of of the new value to be written under the `key`
    /// * `key` - the key to write the new value
    /// calculate the gas to be charged for the writing of the new value.
    /// There is an initial constant value charged for the writing
-   /// If the new value is larger than the old one, then gas is charged for each new byte.
-   /// If the new value is smaller than the old one, then the charge for each saved byte
-   /// is subtracted from the initial value until 0.
-    fn calculate_gas_for_writing(&mut self, new_value_len: u32, key: &str) -> Result<u64> {
-        let mut result = 0;
-        let val = &self.post_execution_state.json[key];
+   /// If the new value is larger than the old one, then gas is charged for the new bytes.
+   /// If the new value is smaller than the old one, then the gas is returned for the removed bytes.
+    fn calculate_gas_for_writing(&mut self, new_value_len: u64, key: &str) -> Result<u64> {
+        let mut result = Some(0);
+        let val = self.post_execution_state.json[key].clone();
```

Comment:
Does it need to be cloned? I can't seem to see where it's used not by reference.

Reply:
self is used by reference